src/medley/core.cljc

```diff
@@ -357,6 +357,64 @@
   ([pred coll]
    (rest (drop-while (complement pred) coll))))
 
+(defn partition-after
+  "Returns a lazy sequence of partitions, splitting after `(pred item)` returns
+  true. Returns a transducer when no collection is provided."
+  ([pred]
+   (fn [rf]
+     (let [v (volatile! [])
+           split? (volatile! false)]
+       (fn
+         ([] (rf))
+         ([result]
+          (rf (if-some [v (not-empty @v)]
+                (unreduced (rf result v))
+                result)))
+         ([result x]
+          (if @split?
+            (let [result (rf result @v)]
+              (vreset! v (if (reduced? result) [] [x]))
+              (when-not (pred x) (vreset! split? false))
+              result)
+            (do (when (pred x) (vreset! split? true))
+                (vswap! v conj x)
+                result)))))))
+  ([pred coll]
+   (lazy-seq
+    (when-let [s (seq coll)]
+      (let [run (take-upto pred s)]
+        (cons run (partition-after pred (lazy-seq (drop (count run) s)))))))))
+
+(defn partition-before
+  "Returns a lazy sequence of partitions, splitting before `(pred item)` returns
+  true. Returns a transducer when no collection is provided."
+  ([pred]
+   (fn [rf]
+     (let [v (volatile! [])]
+       (fn
+         ([] (rf))
+         ([result]
+          (rf (if-some [v (not-empty @v)]
+                (unreduced (rf result v))
+                result)))
+         ([result input]
+          (let [dv @v]
+            (if (and (pred input) (seq dv))
+              (let [result (rf result dv)]
+                (vreset! v (if (reduced? result) [] [input]))
+                result)
+              (do (vswap! v conj input)
+                  result))))))))
+  ([pred coll]
+   (let [cp (complement pred)]
+     ((fn part [coll]
+        (lazy-seq
+         (when-let [s (seq coll)]
+           (let [run (or (seq (take-while cp s))
+                         (cons (first s) (take-while cp (rest s))))]
+             (cons run (part (lazy-seq (drop (count run) s))))))))
```

Comment:
I'm not sure I like the `drop` and `count`, as it means that we need to run through the sequence several times unnecessarily.

Reply:
Yeah I see what you mean, but I'm not sure there is an easy way to determine how much to drop without knowing how many we took. fwiw, this is the approach taken by clojure.core https://github.com/clojure/clojure/blob/clojure-1.10.1/src/clj/clojure/core.clj#L7201

Reply:
Let's keep this for now then, since at least we won't be slower than core, and we can investigate ways of making this more performant later.